src/electron/process_manager.ts

```
@@ -319,53 +285,63 @@ class ChildProcessHelper {
+  constructor(private config: cordova.plugins.outline.ServerConfig) {
     super(pathToEmbeddedBinary('go-tun2socks', 'tun2socks'));
   }
 
-  start(isUdpEnabled: boolean) {
-    // ./tun2socks \
-    //   -tunName "tap0901:outline-tap0:10.0.85.2:10.0.85.0:255.255.255.0" \
+  async start(checkConnectivity: boolean) {
+    // ./tun2socks.exe \
+    //   -tunName outline-tap0" \
     //   -tunAddr 10.0.85.2 -tunGw 10.0.85.1 -tunMask 255.255.255.0 \
-    //   -proxyServer 127.0.0.1:1080 -disableDNSCache -dnsFallback
+    //   -proxyHost 127.0.0.1 -proxyPort 1080 -proxyPassword mypassword \
+    //   -proxyCipher chacha20-ietf-poly1035 -checkConnectivity
     const args: string[] = [];
-    args.push(
-        '-tunName',
-        isWindows ? `tap0901:${TUN2SOCKS_TAP_DEVICE_NAME}:${TUN2SOCKS_TAP_DEVICE_IP}:${
-                        TUN2SOCKS_TAP_DEVICE_NETWORK}:${TUN2SOCKS_VIRTUAL_ROUTER_NETMASK}` :
-                    TUN2SOCKS_TAP_DEVICE_NAME);
+    args.push('-tunName', TUN2SOCKS_TAP_DEVICE_NAME);
     args.push('-tunAddr', TUN2SOCKS_TAP_DEVICE_IP);
     args.push('-tunGw', TUN2SOCKS_VIRTUAL_ROUTER_IP);
     args.push('-tunMask', TUN2SOCKS_VIRTUAL_ROUTER_NETMASK);
-    args.push('-tunDns', `${DNS_RESOLVER_IP_PRIMARY},${DNS_RESOLVER_IP_SECONDARY}`);
-    args.push('-proxyServer', `${this.proxyAddress}:${this.proxyPort}`);
-    args.push('-loglevel', 'error');
-    args.push('-disableDNSCache');
-    if (!isUdpEnabled) {
-      args.push('-dnsFallback');
+    args.push('-tunDNS', DNS_RESOLVERS.join(','));
+    args.push('-proxyHost', this.config.host || '');
+    args.push('-proxyPort', `${this.config.port}`);
+    args.push('-proxyPassword', this.config.password || '');
+    args.push('-proxyCipher', this.config.method || '');
+    args.push('-logLevel', 'info');
+    if (checkConnectivity) {
+      args.push('-checkConnectivity');
     }
-
     this.launch(args);
+
+    const exitListener = this.exitListener;
```

I think this pattern of replacing the listener creates a lot of potential confusion.  For example, the `onExit` setter only works when called before `start`: subsequent changes are silently ignored.

I wonder if this could be simplified by removing `ChildProcessHelper`, now that there's only one subclass.  Alternatively, `Tun2socks` could own a `ChildProcessHelper` instead of subclassing it, to avoid having to change its own public fields.

---

Per our offline conversation, I have encapsulated tun2socks' restart behavior in the `Tun2socks` class. To this end, I followed your suggestion to have `Tun2socks`own a `ChildProcessHelper`. PTAL: b946211.